### PR TITLE
[functorch] add new ensembling api, demonstrate in example

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88850
* #89213

### CODE REVIEW

File: functorch/examples/ensembling/parallel_train.py [hunk 1]
```diff
@@ -68,27 +68,22 @@ def forward(self, x):
...
-# NB: this mechanism doesn't exist in PyTorch today, but we want it to:
-# https://github.com/pytorch/pytorch/issues/49171
-func_model, weights = make_functional(MLPClassifier().to(DEVICE))
-
+model = MLPClassifier().to(DEVICE)
 
 def train_step_fn(weights, batch, targets, lr=0.2):
     def compute_loss(weights, batch, targets):
-        output = func_model(weights, batch)
+        output = torch.nn.utils.stateless.functional_call(model, weights, batch)
```
Comment (zou3519) [inline functorch/examples/ensembling/parallel_train.py:75]: Why isn't this `torch.nn.utils.stateless.functional_call(model, weights, (batch,))` ?

Comment (samdow) [reply functorch/examples/ensembling/parallel_train.py:75]: Great question! In looking into it, it looks like the functional_call docs are different than the behavior, which assumes that `args` is a single argument if it's not a tuple. I'll make another commit updating those docs since it clearly confused me and updating the subsequent commits here to not wrap into a tuple

File: test/functorch/test_eager_transforms.py [hunk 2]
```diff
@@ -3380,12 +3381,14 @@ def compute_loss(weights, batch, targets):
...
-                grad_weights = torch.autograd.grad(loss, weights)
-
-            new_weights = []
-            with torch.no_grad():
-                for grad_weight, weight in zip(grad_weights, weights):
-                    new_weights.append(weight - grad_weight * lr)
+                list_weights = weights if mechanism == "make_functional" else list(weights.values())
+                grad_weights = torch.autograd.grad(loss, list_weights)
+                grad_weights = grad_weights if mechanism == "make_functional" else dict(
+                    zip(weights.keys(), grad_weights))
```
Comment (zou3519) [inline test/functorch/test_eager_transforms.py:3387]: nit: pytrees save us from the if then else:
```
flat_weights, spec = tree_flatten(weights)
flat_grad_weights = torch.autograd.grad(loss, list_weights)
grad_weights = tree_unflatten(flat_grad_weights, spec)
```

File: test/functorch/test_eager_transforms.py [hunk 3]
```diff
@@ -3394,13 +3397,16 @@ def unpack(train_result):
...
+                return combine_state_for_ensemble(models)[1]
+            else:
+                return combine_weights_for_ensemble(models)[0]
 
         def slice_weights(batched_weights, index):
-            return tuple(weight[index].detach().requires_grad_() for weight in batched_weights)
+            if mechanism == "make_functional":
+                return tuple(weight[index].detach().requires_grad_() for weight in batched_weights)
+            else:
+                return {k: weight[index].detach().requires_grad_() for k, weight in batched_weights.items()}
```
Comment (zou3519) [inline test/functorch/test_eager_transforms.py:3409]: could also just tree_map instead of if-then-else

File: test/functorch/test_eager_transforms.py [hunk 4]
```diff
@@ -3456,10 +3463,9 @@ def compute_loss(weights, batch, targets):
...
             grad_weights, loss = grad_and_value(compute_loss)(weights, batch, targets)
-            new_weights = []
-            with torch.no_grad():
-                for grad_weight, weight in zip(grad_weights, weights):
-                    new_weights.append(weight - grad_weight * lr)
+            new_weights = self._update_params(weights, grad_weights, lr, mechanism)
+            if mechanism != "make_functional":
+                new_weights = list(new_weights.values())
             # NB: return looks weird because torch.vmap must return Tensors
             return (loss, *new_weights)
```
Comment (zou3519) [inline test/functorch/test_eager_transforms.py:3470]: (no action required) this comment looks outdated, right?

File: torch/_functorch/make_functional.py [hunk 5]
```diff
@@ -497,6 +497,54 @@ def combine_state_for_ensemble(models):
...
+        assert output.shape == (num_models, batch_size, out_features)
+
+    .. warning::
+        All of the modules being stacked together must be the same (except for
+        the values of their parameters/buffers). For example, they should be in the
+        same mode (training vs eval).
+    """
+    if len(models) == 0:
+        raise RuntimeError('combine_state_for_ensemble: Expected at least one model, got 0.')
+    if not (all(m.training for m in models) or all(not m.training for m in models)):
```
Comment (zou3519) [inline torch/_functorch/make_functional.py:533]: I don't see tests in test_eager_transforms for the error cases here. Would be good to test for the new API.

**Review by zou3519 — COMMENTED** [review]
Directionally this looks good to me and the code looks correct. We should bikeshed the name of the API before this goes in

File: torch/_functorch/make_functional.py [hunk 6]
```diff
@@ -497,6 +497,54 @@ def combine_state_for_ensemble(models):
     return funcs[0], params, buffers
 
 
+def combine_weights_for_ensemble(models):
+    """combine_weights_for_ensemble(models) -> params, buffers
```
[the remark this answers is not in the document]
Comment (samdow) [reply torch/_functorch/make_functional.py:501]: > This includes params and buffers, and "weights" usually just means "params", so shouldn't the name for this be combine_state_for_ensemble? (lol that is what the old one was called)

Yeah in general I was wondering if that was going to make it confusing for users moving from `functorch` to `torch.func` since the signature is going to be different (no function returned). Since IMO if I'm doing that change, I'm doing CMD + F of "functorch" => "torch.func" and then cleaning up the functions that don't exist (mostly make_functional and depending on this combine_state_for_ensemble)

Mostly I was thinking that having the function not exist under the same name as in functorch would make it easier for an existing user to see the changes, but happy to change it back to combine_state_for_ensemble

[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_functorch/make_functional.py:501]: > Yeah in general I was wondering if that was going to make it confusing for users moving from functorch to torch.func since the signature is going to be different (no function returned).

That's a good point. My problem with the name `combine_weights_for_ensemble` is that the implication is that it only combines parameters and does nothing with the buffers

Should we try to come up with another name for this?

[the remark this answers is not in the document]
Comment (samdow) [reply torch/_functorch/make_functional.py:501]: > Should we try to come up with another name for this?

Sure! Spitballing some ideas here:
- combine_params_and_buffers_for_ensemble (maybe too wordy)
- ensembled_state (maybe too close, but might at least be different enough for someone to look at the signature? similar to make_functional => functional_call)
- ensembled_params_and_buffers
- concat_state_for_ensemble (also maybe too close but accurately describes that we are only concating)

File: functorch/examples/ensembling/parallel_train.py [hunk 7]
```diff
@@ -98,7 +93,7 @@ def compute_loss(weights, batch, targets):
 def step4():
     global weights
     for i in range(2000):
-        loss, weights = train_step_fn(weights, points, labels)
+        loss, weights = train_step_fn({k: v for k, v in model.named_parameters()}, points, labels)
```
[the remark this answers is not in the document]
Comment (samdow) [reply functorch/examples/ensembling/parallel_train.py]: Totally works--I just didn't realize until part way through writing these tests. Updated this and previous PR so the only thing that follows this pattern is the detach test. Everything else uses `dict(model.named_parameters())`

File: test/functorch/test_eager_transforms.py (hunk 2 above)
Comment (samdow) [reply test/functorch/test_eager_transforms.py]: Thanks for spotting this! I updated the tests here and in the previous PR to do this

File: test/functorch/test_eager_transforms.py (hunk 4 above)
Comment (samdow) [reply test/functorch/test_eager_transforms.py:3478]: Yep changed this and made it work with pytrees

File: torch/_functorch/make_functional.py [hunk 8]
```diff
@@ -497,6 +497,54 @@ def combine_state_for_ensemble(models):
...
+    .. warning::
+        All of the modules being stacked together must be the same (except for
+        the values of their parameters/buffers). For example, they should be in the
+        same mode (training vs eval).
+    """
+    if len(models) == 0:
+        raise RuntimeError('combine_state_for_ensemble: Expected at least one model, got 0.')
+    if not (all(m.training for m in models) or all(not m.training for m in models)):
+        raise RuntimeError('combine_state_for_ensemble: Expected all models to '
+                           'have the same training/eval mode.')
```
[the remark this answers is not in the document]
Comment (samdow) [reply torch/_functorch/make_functional.py]: Fixed but leaving this open as a reminder to myself once the above bikeshedding is done

File: torch/_functorch/make_functional.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_functorch/make_functional.py:501]: some more ideas after reading yours:
- `stack_module_state` (NB: it's a torch.stack more than a torch.cat)
- `ensembled_module_state` (makes it clearer that it operates on modules, and not e.g. optimizers)

I agree that `combine_params_and_buffers_for_ensemble` is wordy, I could be happy with any of the other names

[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_functorch/make_functional.py:501]: Hmm, so an "ensemble of modules" doesn't necessarily need to have all of its parameters/buffers stacked together. The for-loop approach is still a valid approach. So having stack or concat in the API name might help

[the remark this answers is not in the document]
Comment (samdow) [reply torch/_functorch/make_functional.py:501]: `stack_module_state` makes sense to me! I'll change it

File: torch/_functorch/make_functional.py (hunk 5 above)
Comment (samdow) [reply torch/_functorch/make_functional.py]: Added tests!

File: torch/_functorch/functional_call.py [hunk 9]
```diff
@@ -77,3 +77,52 @@ def functional_call(
         parameters_and_buffers = {k: v for d in parameter_and_buffer_dicts for k, v in d.items()}
 
     return nn.utils.stateless.functional_call(module, parameters_and_buffers, args, kwargs)
+
+
+@exposed_in("torch.func")
+def stack_ensembled_state(models):
+    """stack_ensembled_state(models) -> params, buffers
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:84]: Sorry just to check -- in the comment thread you said you were going with `stack_module_state`, but the API in this PR is `stack_ensembled_state`. Is this intentional?

"stack_ensembled_state" sounds a bit weird to me -- what is an "ensembled state", and what does it mean to "stack" an "ensembled state"? (I can be convinced, this was just my initial reaction)

File: torch/_functorch/functional_call.py [hunk 10]
```diff
@@ -77,3 +77,52 @@ def functional_call(
...
+    .. code-block:: python
+
+        num_models = 5
+        batch_size = 64
+        in_features, out_features = 3, 3
+        models = [torch.nn.Linear(in_features, out_features) for i in range(num_models)]
+        data = torch.randn(batch_size, 3)
+
+        def wrapper(params, buffers, data):
+            return functorch.functional_call(model[0], (params, buffers), data)
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:102]: torch.func.functional_call

File: torch/_functorch/functional_call.py [hunk 11]
```diff
@@ -77,3 +77,52 @@ def functional_call(
...
+    """
+    if len(models) == 0:
+        raise RuntimeError('stack_ensembled_state: Expected at least one model, got 0.')
+    if not (all(m.training for m in models) or all(not m.training for m in models)):
+        raise RuntimeError('stack_ensembled_state: Expected all models to '
+                           'have the same training/eval mode.')
+    model0_typ = type(models[0])
+    if not all(type(m) == model0_typ for m in models):
+        raise RuntimeError('stack_ensembled_state: Expected all models to '
+                           'be of the same class.')
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:122]: I might not be looking in the right places, but, are there tests that check these errors get hit?

**Review by zou3519 — COMMENTED** [review]
I had two last two points, then this should be good to go!
- My first reaction at `stack_ensembled_state` was some confusion, but I can be convinced
- Are there tests for the error cases?

File: torch/_functorch/functional_call.py (hunk 9 above)
Comment (samdow) [reply torch/_functorch/functional_call.py]: Ahh good catch--this was unintentional. Changed it to `stack_module_state`

File: torch/_functorch/functional_call.py (hunk 11 above)
Comment (samdow) [reply torch/_functorch/functional_call.py]: Somehow these test got added to a different commit so thanks for calling it out. They're now in `test_stack_module_state_error` in `test_eager_transforms.py`

File: torch/_functorch/functional_call.py (hunk 10 above)
Comment (samdow) [reply torch/_functorch/functional_call.py:128]: Fixed

**Comment by zou3519** [comment]
Looks like there are some erroneous submodule changes

File: torch/_functorch/functional_call.py [hunk 12]
```diff
@@ -103,3 +103,52 @@ def compute_loss(params, x, t):
         parameters_and_buffers = {k: v for d in parameter_and_buffer_dicts for k, v in d.items()}
 
     return nn.utils.stateless.functional_call(module, parameters_and_buffers, args, kwargs)
+
+
+@exposed_in("torch.func")
+def stack_module_state(models):
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:109]: nit: add type hints:

stack_module_state(models: List[nn.Module]) -> Tuple[Dict[str, Any], Dict[str, Any]]

File: torch/_functorch/functional_call.py [hunk 13]
```diff
@@ -103,3 +103,52 @@ def compute_loss(params, x, t):
...
+    """stack_module_state(models) -> params, buffers
+
+    Prepares a list of torch.nn.Modules for ensembling with :func:`vmap`.
+
+    Given a list of ``M`` ``nn.Modules`` of the same class, returns two dictionaries
+    that stack all of their parameters and buffers together, indexed by name.
```
Comment (zou3519) [inline torch/_functorch/functional_call.py:115]: nit: maybe add an example with submodules that prints out the dictionaries so the user can see the convention of the name strings?

**Review by zou3519 — APPROVED** [review]
LGTM

File: torch/_functorch/functional_call.py (hunk 12 above)
Comment (samdow) [reply torch/_functorch/functional_call.py]: added!

File: torch/_functorch/functional_call.py (hunk 13 above)
Comment (samdow) [reply torch/_functorch/functional_call.py:115]: added!
